lib/torrent.js

```diff
@@ -638,24 +639,71 @@ export default class Torrent extends EventEmitter {
     })
   }
 
-  _verifyPieces (cb) {
-    parallelLimit(this.pieces.map((piece, index) => cb => {
+  _verifyPiece (index, cb) {
+    if (this.destroyed) return cb(new Error('torrent is destroyed'))
+
+    const getOpts = {}
+    // Specify length for the last piece in case it is zero-padded
+    if (index === this.pieces.length - 1) {
+      getOpts.length = this.lastPieceLength
+    }
+    this.store.get(index, getOpts, async (err, buf) => {
       if (this.destroyed) return cb(new Error('torrent is destroyed'))
 
-      const getOpts = {}
-      // Specify length for the last piece in case it is zero-padded
-      if (index === this.pieces.length - 1) {
-        getOpts.length = this.lastPieceLength
-      }
-      this.store.get(index, getOpts, async (err, buf) => {
-        if (this.destroyed) return cb(new Error('torrent is destroyed'))
+      if (err) return queueMicrotask(() => cb(null, false)) // ignore error
+
+      const hex = await hash(buf, 'hex')
+      if (this.destroyed) return cb(new Error('torrent is destroyed'))
 
-        if (err) return queueMicrotask(() => cb(null)) // ignore error
+      cb(null, hex === this._hashes[index])
+    })
+  }
 
-        const hex = await hash(buf, 'hex')
-        if (this.destroyed) return cb(new Error('torrent is destroyed'))
+  _verifyPiecesUsingBitfield (cb) {
+    const piecesToCheck = new Set()
+    this.bitfield.buffer = new Uint8Array(this._startupBitfield)
+    this.files.forEach(file => {
+      let checkFile = true
+      for (let i = file._startPiece; i <= file._endPiece; ++i) {
+        if (this.bitfield.get(i)) {
+          if (checkFile) {
+            piecesToCheck.add(i)
+            checkFile = false
+          }
+          this._debug('piece verified %s', i)
+          this._markVerified(i)
+        } else {
+          this._markUnverified(i)
+          this._debug('piece invalid %s', i)
+        }
+      }
+    })
+    parallelLimit([...piecesToCheck].map((index) => cb => {
+      this._verifyPiece(index, (err, isVerified) => {
+        if (err) return cb(err)
+        if (!isVerified) {
+          this._startupBitfield = null
+          return cb(new Error('piece is invalid, force recheck'))
```

Comment:
```suggestion
          return cb(index)
        }
        cb(null)
      })
    }), FILESYSTEM_CONCURRENCY, (err) => {
      if (err) {
        const filesToRecheck = []
        for (const index of err) {
          for (const file of this.files) {
            if (file.includes(err)) {
              filesToRecheck.push(file)
            }
          }
        }
        return rescanSpecificFiles(filesToRecheck, cb)
      }
      cb(null)
```

I think something like this *might* work?

Reply:
actually something along the lines of
```js
let checkFile = 2
if (checkFile) {
  piecesToCheck.add(i)
  if (checkFile-- === 1) {
    piecesToCheck.delete(i - 1)
  }
}
```
could probably do the job

Reply:
I rewrote it differently:

```js
const piecesToCheck = new Set()
    const piecesToFilesMap = new Map()
    this.files.forEach(file => {
      const previouslyVerifiedFilePieces = []
      for (let i = file._startPiece; i <= file._endPiece; ++i) {
        if (this.bitfield.get(i)) {
          previouslyVerifiedFilePieces.push(i)
          if (!piecesToFilesMap.has(i)) {
            piecesToFilesMap.set(i, [])
          }
          piecesToFilesMap.get(i).push(file)
          this.emit('verified', i)
        } else {
          this._markUnverified(i)
          this._debug('piece invalid %s', i)
        }
      }
      const nonOverlappingPiece = previouslyVerifiedFilePieces.find(piece => piecesToFilesMap.get(piece).length === 1)
      if (nonOverlappingPiece) {
        piecesToCheck.add(nonOverlappingPiece)
      } else if (previouslyVerifiedFilePieces.length >= 1) {
        piecesToCheck.add(previouslyVerifiedFilePieces[0])
      }
    })
```

it is more complex but it prioritizes truly non-overlapping pieces.

Also, I don't see a big problem with overlapping pieces, it is verifying them using the hashes and not just their existence, so if a piece has the correct hash and spans across multiple files all those files are correct.

Considering the most probable use ( start and stop of an application that uses webtorrent and re-loads partially downloaded torrents ) the chances of someone tampering with the contents of those files are close to 0, and what truly can happen is that the download folder will be deleted